### ISSUE TITLE
Idea for the Recipe Book

### ISSUE THREAD

**Issue by RedRain0o0** [hammy3502/immersive-mc#164]
In the current state, there is no Immersive way to open or use the Recipe Book, and I have a suggestion.
This would be VR only, as I see no real way to make this work in 2d.

In VR, approaching an Immersive that normally has a Recipe Book will put one in your Off-Hand, but only visually, your actual item will still be there, it'll just be invisible(If it's a shield, it'll be disabled)

There will be 2 sections, the recipe icons, and the recipes themselves. The icons would be at the start, like a table of contents/shortcut
The table of contents will have the normal grid of recipes displayed on each page, and however many pages are needed.
On the side will be the normal Recipe tabs, and you just pull them to switch to that section.
To flip pages, you bring your hand to one end and swipe, like your turning the pages of a book.
Touching a recipe icon will flip to it's recipe page at the back of the book

You would be able to have your bag open at the same time
If there are more than 1 Recipe Book Immersive, punching an Immersive will swap the book to that ones Book.
You would need to be within a ~3 block radius of the Immersive for the Book to appear.

**Comment by hammy275** [comment]
Hello! I want to take the moment to thank you for the suggestion, I do greatly appreciate suggestions, and if you have any more, I would be interested in hearing them!

That being said, I'm going to respectfully have to decline implementing this, either for the time being, or as a whole. My reasons for so are listed below:

- To my admittedly very limited knowledge, most players playing Minecraft in VR already tend to know the recipes for the game (as they've already played Minecraft for a while), and thus don't need the recipe book. Admittedly though, I'm sure it would help with placing items into the crafting table much more quickly.
- As much as I want to have it in place, I currently don't have the infrastructure in place code-wise for replacing items as you described. I may have it in place in the future, as I would like users to be able to hammer items together in an anvil, but at this time, it is currently not in place, and will require a lot of work to get in place.


Tl;dr: It's a lot for a relatively small subset of users.

A lot of the components that would make up the recipe book I see myself possibly implementing in the future (invisible items with the anvil hammer and a book with pages for an immersive written book if I ever decide to implement that), so if all those pieces are implemented at some point, I may just put them together to make the recipe book as well. Until then though, I'm going to respectfully pass on adding this.

Thank you again for the suggestion though, and if you have any other suggestions, I'd love to hear them! :)

**Comment by hammy275** [comment]
The work in ImmersiveMC's codebase has been basically done for this, and I see this to be really useful, not necessarily for learning recipes, but for crafting items more quickly, similarly to the JEI-family of mods.

**Comment by Adidiii** [comment]
Another idea would be to have the book on the side of a workbench, and picking it up (optionally, it wouldn't be forced into your hand) would have it so it would be readable similarly to immersive books.

**Comment by hammy275** [comment]
As suggested in Discord, this should also be included on the bag for 2x2 crafting.

**Comment by Steveplays28** [comment]
I would like to add to this: it might be cool to have (optional) EMI integration. EMI has an API that allows rendering (parts of) the recipe viewer screen, which could be put on a book in a similar way to NotEnoughAnimations' in-world books.

Maybe the list of items could be on the left side of the book and the recipe(s) could be on the right side of the book.
This would function similarly to the recipe book idea above but with EMI's features, including more complex recipe and tag support.

**Comment by hammy275** [comment]
For getting recipes and performing auto-filling, based on the suggestion above, I think it makes sense to attempt to use recipe mods, then fallback to the recipe book if all else fails. This way, we may have access to modded recipes on modded blocks more easily and likely have better filling code than what I could write.

This means I'll need 4 implementations:
- EMI
- REI
- JEI
- Recipe book (the null implementation/fallback/default)

I'm pretty sure at least one of REI and/or EMI have extra integration if JEI is installed, so I need to make sure I prioritize different mods correctly here. Furthermore, everything except the recipe book stuff, of course, should be wrapped in a `CompatModule`.

EDIT: As a bonus, this means for JEI/REI/EMI, we can offer integration into the favorites system! I won't be building my own for recipe books. In short, I'm not keen to write a part of JEI/REI/EMI just to make this feature happen.

**Comment by Steveplays28** [comment]
That would be very cool, exactly what I imagined with that proposal!
Thank you for looking into this

**Comment by hammy275** [comment]
Needed to think for a while on how to handle item organization in the book so items can be quickly accessed without reducing usability or impacting other languages.

It'll be a couple "clicks":
- First click is to select the category the item is in. The category is based on the creative mode tabs, as mods add their own creative mode tab, but the recipe book only has 5 total categories.
- Next click is the actual item, of course only showing items you can craft.

Any categories containing 0 items to craft will intentionally still be shown for muscle memory reasons, and if you're required to have the recipe unlocked to craft something (I forget the name of the gamerule), we'll hide anything you don't have unlocked.*

Categories and items are displayed with one per row, with the item being rendered to the left of the page. Left-click to put the item in as input (puts in as many items as possible).

That aside, I'm realizing supporting a fallback, JEI, REI, and EMI is likely not feasible. I think just a built-in "vanilla" option would be better for the sake of maintainability.

*: Newer Minecraft versions outright don't sync unlocked recipes. Between syncing all recipes over or just having people unlock them, I lean towards the latter for simplicity. Most unlock conditions on Minecraft's part are pretty reasonable, anyways.

**Comment by hammy275** [comment]
Definitely only want to do a fallback. Both due to the feasibility of maintaining 4 implementations and because it appears like either the APIs don't support what I'm looking for and/or depend on a screen being involved. Even if they do though, 4 implementations is too much 😅 .

I won't abstract this for other crafting types, just crafting tables (and the Tinkers' Construct Crafting Station and Visual Workbench).

Code for filling will run entirely on the client to massively reduce complexity. This _would_ be a massive pain, were it not for the recipe book already handling all the complexity! In short, we can (every tick, since we can't monitor for inventory changes): go to the recipe book on the client, get all the recipes it says we can craft, then populate the book accordingly. See `RecipeCollection#hasCraftable()` if we can craft it, and `ClientRecipeBook#getCollections()` to get all the recipes.

From here, the client picks out all the slots containing valid items, then sends them to the server as swap packet(s).
